### PR TITLE
Add 810 & 855 for v4010

### PR DESCRIPTION
**Pull request by kelnadim**

### CODE REVIEW

**Review by kputnam — COMMENTED** [review]
This looks good. There are two general issues that need to be addressed before merging:

* Match the whitespace and vertical alignment style used in the project. I know it can be a pain if your editor is configured to automatically reformat what you type.
* Add at least one valid X12 fixture file for each transaction, in `spec/fixtures`, following the organization convention used there. You can replace any sensitive data with generic values.

You won't need to write any specs yourself, as specs will automatically be generated that ensure your fixture files are parsed correctly. However, the auto-generated specs cannot test that your transaction definition is correct. To do that, you will need to write specs similar to [this one](https://github.com/irobayna/stupidedi/blob/master/spec/lib/stupidedi/transaction_sets/005010/X212-HR276_spec.rb).

File: lib/stupidedi/transaction_sets/004010/implementations/IN810.rb [hunk 1]
```diff
@@ -0,0 +1,191 @@
+module Stupidedi
+  module TransactionSets
+    module FortyTen
+      module Implementations
+        b = Builder
+        d = Schema
+        r = SegmentReqs
+        e = ElementReqs
+        s = SegmentDefs
+        IN810 = b.build("IN", "810", "Invoice",
```
Comment (kputnam) [inline lib/stupidedi/transaction_sets/004010/implementations/IN810.rb:10]: Continued lines should only be indented two spaces more than the previous line

File: lib/stupidedi/transaction_sets/004010/implementations/IN810.rb [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                                                                       b::Element(e::Required,    "Entity Identifier Code"),
+                                                                       b::Element(e::Situational, "Name"),
+                                                                       b::Element(e::Situational, "Identification Code Qualifier"),
+                                                                       b::Element(e::Situational, "Identification Code"),
+                                                                       b::Element(e::Situational, "Code describing entity relationship"),
+                                                                       b::Element(e::Situational, "Code identifying an organizational entity, a physical location, property or an individual")),
+                                                            b::Segment(120, s:: N2, "Address Information", r::Situational, d::RepeatCount.bounded(2),
+                                                                       b::Element(e::Situational,    "Name"),
```
Comment (kputnam) [inline lib/stupidedi/transaction_sets/004010/implementations/IN810.rb:40]: The beginning of each argument should be aligned vertically with the previous line.

File: lib/stupidedi/transaction_sets/004010/implementations/IN810.rb [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                                                      b::Element(e::Situational, "Terms Net Days"),
+                                                      b::Element(e::Situational, "Terms Discount Amount"),
+                                                      b::Element(e::Situational, "Terms Deferred Due Date"),
+                                                      b::Element(e::Situational, "Deferred Amount Due"),
+                                                      b::Element(e::Situational, "Percent of Invoice Payable"),
+                                                      b::Element(e::Situational, "Description"),
+                                                      b::Element(e::Situational, "Day of Month"),
+                                                      b::Element(e::Situational, "Payment Method Code"),
+                                                      b::Element(e::Situational, "Percent"))
+                        ), #end TableDef HEADING
```
Comment (kputnam) [inline lib/stupidedi/transaction_sets/004010/implementations/IN810.rb:68]: Closing parentheses should not be on their own line. You can leave the `# end XYZ` markers. In cases where there are several consecutive closing parentheses, they need to be written side-by-side; multiple end markers would need to be joined like so: `...))) # end A, B, C`

File: lib/stupidedi/transaction_sets/004010/implementations/IN810.rb [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                                                                                        b::Element(e::Situational, "Agency Qualifier Code"),
+                                                                                        b::Element(e::Situational, "Product Description Code"),
+                                                                                        b::Element(e::Situational, "Description"),
+                                                                                        b::Element(e::Situational, "Surface/Layer/Position Code"),
+                                                                                        b::Element(e::Situational, "Source Subqualifier"),
+                                                                                        b::Element(e::Situational, "Yes/No Condition or Response Code"),
+                                                                                        b::Element(e::Situational, "Language Code")
+                                                                             )
```
Comment (kputnam) [inline lib/stupidedi/transaction_sets/004010/implementations/IN810.rb:122]: Consecutive closing parentheses should be written side-by-side (and should not be on their own line). Eg, `b::Element(..., ...)))`

File: lib/stupidedi/versions/004010/element_defs.rb [hunk 5]
```diff
@@ -9461,6 +9466,71 @@ module ElementDefs
           s::CodeList.external("559"))
         E1715 = t::ID.new(:E1715, "Country Subdivision Code"             , 1, 3,
           s::CodeList.external("5"))
+        E1716  = t::AN.new(:E1716 , "Change Order Sequence Number"       , 1, 8)
+        E1717  = t::ID.new(:E1717 , "Allowance or Charge Indicator"      , 1, 1,
+                           s::CodeList.build(
+                               "A"   => "Allowance",
```
Comment (kputnam) [inline lib/stupidedi/versions/004010/element_defs.rb:9472]: The `=>` tokens line up vertically, but you don't need the extra preceding spaces.

File: lib/stupidedi/versions/004010/element_defs.rb [hunk 6]
```diff
@@ -9461,6 +9466,71 @@ module ElementDefs
...
+                               "C"   => "Charge",
+                               "N"   => "No Allowance or Charge"))
+        E1718  = t::ID.new(:E1718 , "Allowance or Charge Indicator"      , 4, 4,
+                           s::CodeList.build(
+                               "A310"   => "Air Express Charge",
```
Comment (kputnam) [inline lib/stupidedi/versions/004010/element_defs.rb:9477]: Too many spaces before `=>` here too.

File: lib/stupidedi/versions/004010/segment_defs/ACK.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+module Stupidedi
+  module Versions
+    module FortyTen
+      module SegmentDefs
+        s = Schema
+        e = ElementDefs
+        r = ElementReqs
+
+        ACK = s::SegmentDef.build(:ACK, "Line Item Acknowledgment",
+          "To start acknowledgement of a line item",
```
Comment (kputnam) [inline lib/stupidedi/versions/004010/segment_defs/ACK.rb:11]: This file is formatted correctly, except one issue at the bottom

File: lib/stupidedi/versions/004010/segment_defs/ACK.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+          e::E234 .simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+          e::E235 .simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+          e::E234 .simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+          e::E235 .simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+          e::E234 .simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+          e::E235 .simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+          e::E234 .simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+          e::E235 .simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+          e::E234 .simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+          e::E1271  .simple_use(r::Optional,  s::RepeatCount.bounded(1)))
```
Comment (kputnam) [inline lib/stupidedi/versions/004010/segment_defs/ACK.rb:40]: The `.` tokens should line up vertically

File: lib/stupidedi/versions/004010/segment_defs/BIG.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+                                  e::E328.simple_use(r::Optional,  s::RepeatCount.bounded(1)),
+                                  e::E1716.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E640.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E353.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E306.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E76.simple_use(r::Optional,   s::RepeatCount.bounded(1)))
+      end
+    end
+  end
+end
```
Comment (kputnam) [inline lib/stupidedi/versions/004010/segment_defs/BIG.rb:26]: Each file should have a trailing blank line. Your other files do have one, but sometimes text editor configuration can mess with these.

File: lib/stupidedi/versions/004010/segment_defs/ITD.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+                                  e::E1741.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E954.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E352.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E1742.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E1743.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E954.simple_use(r::Optional,   s::RepeatCount.bounded(1)))
+      end
+    end
+  end
+end
```
Comment (kputnam) [inline lib/stupidedi/versions/004010/segment_defs/ITD.rb:30]: One more file that needs a trailing blank line

File: lib/stupidedi/versions/004010/segment_defs/SAC.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+                                  e::E1722.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E127.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E1723.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E352.simple_use(r::Optional,   s::RepeatCount.bounded(1)),
+                                  e::E819.simple_use(r::Optional,   s::RepeatCount.bounded(1))
+        )
+      end
+    end
+  end
+end
```
Comment (kputnam) [inline lib/stupidedi/versions/004010/segment_defs/SAC.rb:32]: Another missing blank line. I guess there were more than I thought. These are marked on GitHub with a `⊖` symbol at the end of the file

**Comment by kelnadim** [comment]
@kputnam Thanks for your feedback. I believe I addressed all your concerns.

File: spec/fixtures/004010/IN810/pass/1.edi [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+ST|810|5639689~
+BIG|20211118|00237058|20211102|0691-0014403~
+N1|ST|Store Shipping Address|ZZ|1091581~
+N3|123 address_1|address_2~
+N4|SAVANNAH|GA|314014245~
+ITD||3|||||30~
+IT1||4||70||||EN|9781501331978|GC|01~
+PID|F||||Retail Buying From Basics to Fashion~
+IT1||1||5.66||||EN|9780316154680|GC|02~
+PID|F||||When You Are Engulfed in Flames~
```
Comment (kputnam) [inline spec/fixtures/004010/IN810/pass/1.edi:12]: Haha, this looks like a good read. Thanks for the recommendation!

**Review by kputnam — APPROVED** [review]
Looks great, thanks for the reminder to review!

**Comment by kputnam** [comment]
@irobayna For some reason the CI builds aren't running, but the test suite passed locally. I believe this is ready for release.

**Pull request merged**